test/e2e/upgrade_cluster_test.go

```diff
@@ -0,0 +1,186 @@
+		clusterOpts.AWSPlatform.Zones = zones
+		clusterOpts.InfrastructureAvailabilityPolicy = string(hyperv1.HighlyAvailable)
+		clusterOpts.NodePoolReplicas = 1
+	}
+	clusterOpts.BeforeApply = func(o crclient.Object) {
+		switch obj := o.(type) {
+		case *hyperv1.HostedCluster:
+			if obj.Annotations == nil {
+				obj.Annotations = make(map[string]string)
+			}
+			obj.Annotations[hyperv1.ControlPlaneOperatorV2Annotation] = "true"
+		}
+	}
+
+	if globalOpts.RunUpgradeTestSetup {
+		t.Log("Starting HyperShift Operator upgrade test setup")
+		t.Run("Setup", func(t *testing.T) {
+			setupUpgradeTest(t, ctx, cancel, hcInfoFilePath, clusterOpts, zones)
+		})
+	} else if globalOpts.RunUpgradeTest {
+		t.Log("Starting HyperShift Operator upgrade test")
+		runUpgradeTest(t, hcInfoFilePath, ctx, cancel, clusterOpts, zones)
+	} else {
+		// This test should be triggered intentionally by setting the options above,
+		// otherwise skip it.
+		t.SkipNow()
+	}
+}
+
+func runUpgradeTest(t *testing.T, hcInfoFilePath string, ctx context.Context, cancel context.CancelFunc, clusterOpts e2eutil.PlatformAgnosticOptions, zones []string) {
+	g := NewWithT(t)
+	jsonData, err := os.ReadFile(hcInfoFilePath)
+	g.Expect(err).NotTo(HaveOccurred(), "couldn't write hcinfo file")
+	var preUpgradeInfo PreUpgradeInfo
+	err = json.Unmarshal(jsonData, &preUpgradeInfo)
+	g.Expect(err).NotTo(HaveOccurred(), "couldn't unmarshal hcinfo")
+	t.Logf("Loaded pre-upgrade info: %+v", preUpgradeInfo)
+
+	namespace := preUpgradeInfo.Namespace
+
+	client, err := e2eutil.GetClient()
+	g.Expect(err).NotTo(HaveOccurred(), "couldn't get k8s client")
+
+	hostedCluster := hyperv1.HostedCluster{}
+	if err := client.Get(ctx, crclient.ObjectKey{Namespace: namespace, Name: preUpgradeInfo.Name}, &hostedCluster); err != nil {
+		t.Fatalf("failed to get hosted cluster %s in namespace %s: %v", preUpgradeInfo.Name, namespace, err)
+	}
+	// Now we have what we need to cleanup properly.
+	defer func() {
+		t.Run("Teardown", func(t *testing.T) {
+			e2eutil.TeardownHC(t, ctx, &hostedCluster, client, &clusterOpts, globalOpts.ArtifactDir, false)
+			cancel()
+		})
+	}()
+
+	// Get the newly created default NodePool
+	nodepools := &hyperv1.NodePoolList{}
+	if err := client.List(ctx, nodepools, crclient.InNamespace(namespace)); err != nil {
+		t.Fatalf("failed to list nodepools in namespace %s: %v", namespace, err)
+	}
+	g.Expect(nodepools.Items).ToNot(BeEmpty())
+	g.Expect(len(nodepools.Items)).To(BeEquivalentTo(len(zones)))
+
+	nodePool := nodepools.Items[0]
+
+	g.Expect(nodePool.Generation).To(Equal(preUpgradeInfo.NodePoolGeneration))
+	g.Expect(nodePool.Annotations[nodePoolAnnotationCurrentConfig]).To(Equal(preUpgradeInfo.NodePoolCurrentConfig))
+	g.Expect(nodePool.Annotations[nodePoolAnnotationCurrentConfigVersion]).To(Equal(preUpgradeInfo.NodePoolCurrentConfigVersion))
+
+	// Get the newly created default MachineDeployment
```

Comment:
Why are we doing asserts on MachineDeployement underneath NodePools? It seems like checking the `NodePool` is sufficient.

Reply:
I agree NodePools  should be sufficient in principle. Though I think additionally asserting on the MachineDeployment is a good extra safety check.